core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -652,13 +659,16 @@ private[deploy] class Master(
       val usableWorkers = workers.toArray.filter(_.state == WorkerState.ALIVE)
         .filter(worker => worker.memoryFree >= app.desc.memoryPerExecutorMB &&
           worker.coresFree >= coresPerExecutor.getOrElse(1))
-        .sortBy(_.coresFree).reverse
-      val assignedCores = scheduleExecutorsOnWorkers(app, usableWorkers, spreadOutApps)
+        .sortWith { (l, r) =>
+          (l.coresFree.toFloat / l.cores) < (r.coresFree.toFloat / r.cores)
```

Comment:
Sorting by percent instead of count makes the behavior very different from the prior impl.